### ISSUE TITLE
lagbuffer and TimeControl

### ISSUE THREAD

**Issue by dbosst** [leela-zero/leela-zero#2296]
The following applies to a client not sending constant updates on every move using the GTP command time_left, and does not affect pure byo-yomi time controls.

The current implementation for lag buffer makes LZ often lose on time when using Canadian time control (and for Absolute time too). This is mainly because **LZ does not take the lag into account to correct it's time remaining when a move is made.**

This makes LZ clock get out of sync with the client clock as more and more moves are made. This doesn't affect byo-yomi with zero main time, but it does affect Canadian and Absolute timing control.  Additionally, when LZ calculates the maximum move time it calculates using the lag over an entire period, but when the above is true, the lag really should add up for each move in a period; this results  in moves later in the period get less and less time than they should.

At the moment, with the current implementation the client either has to set a very large lagbuffer, or, try to guess and adjust the lagbuffer based on the number of stones.

**Comment by gcp** [comment]
The lagbuffer logic most definitely assumes that the stated lag is a maximum and that it will get updates from the client via time_left.

**Comment by gcp** [comment]
I'm not really in favor of even trying to fix interfaces that lag and can't update their elapsed time correctly. This seems like it won't ever work correctly.

**Comment by gcp** [comment]
The point of lagbuffer is to deal with lag, i.e. say you're playing KGS with a 10 second Japanese byo yomi, and you get a 1s lag spike at any point during the game, suddenly the engine forfeits on time because it was allocating 10s of time for your move.

That is, it's mean to deal with spikes. It wasn't meant to deal with continuous lag because time_left will adjust that.

If the GUI lags *and* doesn't give time_left updates, bah!

**Comment by Dorus** [comment]
@gcp i agree the client should correct it with `time_left`, and adding a 1 second lag buffer (or whatever the user set it at) is fair.

But doesn't it make sense to also add a (smaller) lag time each move if the client doens't share a new `time_left` number?

**Comment by dbosst** [comment]
> The lagbuffer logic most definitely assumes that the stated lag is a maximum and that it will get updates from the client via time_left.

I see so the client is then required to always send time_left before every genmove. This is a sort of grey area of GTP and I didn't know what was commonly expected. In that case I will close this and the pull request.

**Comment by gcp** [comment]
@dbosst I don't think GTP specifies it, but logically things can't work very well if the GUI lags significantly and never synchronizes the time.

I do think you're on the something that the calculation for Canadian style byo yomi (at least) is not correct because it doesn't account for the fact that there might be lag on every move - the patch you submitted just had a bug, I think.

There MAY very well be a point to accounting for a constant per-move lag, and taking this into account in time calculations (cfr. @Dorus suggestion). Stockfish does this, but then again extremely fast timecontrols are more common in chess. Conversely, having to account for server lag spikes is less commonly an issue due to lag compensation on those servers.

I'll keep this bug open for a bit for discussion.

**Comment by dbosst** [comment]
The reason for me noticing it is I'm working on a fork of Sabaki with clocks, and noticed how finicky Canadian timing is to get LZ to not timeout without the client sending time_left constantly. (To say nothing of the lack of full byoyomi support for time_left in GTP which is a different issue).
